src/level.cpp

```diff
@@ -997,14 +998,250 @@ void Level::GenerateRasterizableData(std::vector<Quadblock>& quadblocks)
 		}
 	}
 	highLODMesh.UpdateMesh(highLODData, (Mesh::VBufDataType::VColor | Mesh::VBufDataType::Normals), Mesh::ShaderSettings::None);
-	m_highLODLevelModel.SetMesh(&highLODMesh);
+	this->m_highLODLevelModel.SetMesh(&highLODMesh);
 
 	lowLODMesh.UpdateMesh(lowLODData, (Mesh::VBufDataType::VColor | Mesh::VBufDataType::Normals), Mesh::ShaderSettings::None);
-	m_lowLODLevelModel.SetMesh(&lowLODMesh);
+	this->m_lowLODLevelModel.SetMesh(&lowLODMesh);
 
 	vertexHighLODMesh.UpdateMesh(vertexHighLODData, (Mesh::VBufDataType::VColor | Mesh::VBufDataType::Normals), Mesh::ShaderSettings::None);
-	m_pointsHighLODLevelModel.SetMesh(&vertexHighLODMesh);
+	this->m_pointsHighLODLevelModel.SetMesh(&vertexHighLODMesh);
 
 	vertexLowLODMesh.UpdateMesh(vertexLowLODData, (Mesh::VBufDataType::VColor | Mesh::VBufDataType::Normals), Mesh::ShaderSettings::None);
-	m_pointsLowLODLevelModel.SetMesh(&vertexLowLODMesh);
+	this->m_pointsLowLODLevelModel.SetMesh(&vertexLowLODMesh);
+}
+
+void Level::GenerateRenderBspData(BSP bsp)
+{
+	static Mesh bspMesh;
+	std::vector<float> bspData;
+
+	auto point = [](const Vertex* verts, int ind, std::vector<float>& data) {
+		data.push_back(verts[ind].m_pos.x);
+		data.push_back(verts[ind].m_pos.y);
+		data.push_back(verts[ind].m_pos.z);
+		data.push_back(verts[ind].m_normal.x);
+		data.push_back(verts[ind].m_normal.y);
+		data.push_back(verts[ind].m_normal.z);
+		Color col = verts[ind].GetColor(true);
+		data.push_back(col.r);
+		data.push_back(col.g);
+		data.push_back(col.b);
+		};
+
+	std::function<void(const BSP*, int)> createGeom = [&createGeom, &point, &bspData](const BSP* b, int depth) {
+		if (GuiRenderSettings::bspTreeMaxDepth < depth)
+			GuiRenderSettings::bspTreeMaxDepth = depth;
+		const BoundingBox& bb = b->GetBoundingBox();
+		Color c = Color(depth * 30.0, 1.0, 1.0);
+		Vertex verts[] = {
+			Vertex(Point(bb.min.x, bb.min.y, bb.min.z, c.rb, c.gb, c.bb)), //---
+			Vertex(Point(bb.min.x, bb.min.y, bb.max.z, c.rb, c.gb, c.bb)), //--+
+			Vertex(Point(bb.min.x, bb.max.y, bb.min.z, c.rb, c.gb, c.bb)), //-+-
+			Vertex(Point(bb.max.x, bb.min.y, bb.min.z, c.rb, c.gb, c.bb)), //+--
+			Vertex(Point(bb.max.x, bb.max.y, bb.min.z, c.rb, c.gb, c.bb)), //++-
+			Vertex(Point(bb.min.x, bb.max.y, bb.max.z, c.rb, c.gb, c.bb)), //-++
+			Vertex(Point(bb.max.x, bb.min.y, bb.max.z, c.rb, c.gb, c.bb)), //+-+
+			Vertex(Point(bb.max.x, bb.max.y, bb.max.z, c.rb, c.gb, c.bb)), //+++
+		};
+		//these normals are octohedral, should technechally be duplicated and vertex normals should probably be for faces.
+		verts[0].m_normal = Vec3(-1.f / 1.44224957031f, -1.f / 1.44224957031f, -1.f / 1.44224957031f);
```

Comment:
Could you put the constant `1.44224957031f` into a `static constexpr` variable? would make the code easier to read and less error prone whenever this constant needs to be used